processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,45 +133,100 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
+    final Integer priorIdex = facts.get(key);
+
+    Aggregator[] aggs;
+
+    if (null != priorIdex) {
+      aggs = atomicGet(priorIdex);
+    } else {
+      aggs = new Aggregator[metrics.length];
+      for (int i = 0; i < metrics.length; i++) {
+        final AggregatorFactory agg = metrics[i];
+        aggs[i] = agg.factorize(
+            makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
+        );
+      }
+      final Integer rowIndex = indexIncrement.getAndIncrement();
+
+      // We always want to have a NULL value to replace
+      if (aggListReference.get().length <= rowIndex) {
+        ensureCapacity(aggListReference.get().length<<1);
+      }
+
+      atomicSet(rowIndex, aggs);
```

Comment:
there is technically no guarantee that the aggList at that point still has the necessary capacity, it could have been filled up already by other threads.